### ISSUE TITLE
Aspect ratio for not so common scales

### ISSUE THREAD

**Issue by ImportanceOfBeingErnest** [matplotlib/matplotlib#13819]
### Bug report

**Bug summary**

Bringing up [this stackoverflow question](https://stackoverflow.com/questions/55438189/matplotlib-symlog-plot-with-non-auto-aspect-ratio) on the issue tracker.

Setting an equal aspect for symlog scales results in an `UnboundLocalError`. This should never happen, independent on how meaningful the notion of "aspect" would be such case.

**Code for reproduction**

```
import matplotlib.pyplot as plt
import numpy as np

dt = 0.01
x = np.arange(-50.0, 50.0, dt)
y = np.arange(0, 100.0, dt)

fig = plt.figure()
ax = fig.add_subplot(111,aspect='equal')

plt.plot(x, np.sin(x / 3.0))
plt.xscale('symlog')
plt.yscale('symlog', linthreshy=0.015)
plt.grid(True)
plt.ylabel('symlog both')

plt.show()
```

**Actual outcome**

```
UnboundLocalError: local variable 'aspect_scale_mode' referenced before assignment
```

**Expected outcome**

Either of the following:

* An error telling that symlog does not support equal aspect.
* A warning of the above and a fallback to `aspect="auto".
* A meaningful plot, in case someone has an idea of what "aspect" could mean in the case of two symlog scales.

**Analysis**

The problem comes from https://github.com/matplotlib/matplotlib/blob/699f35e4d8d424393a7038428abc573d591bd6db/lib/matplotlib/axes/_base.py#L1469-L1485

where there are a lot of cases for which the variable `aspect_scale_mode` isn't defined. I.e. each case that ends up in the `pass` section.
When later accessing `aspect_scale_mode` in 
https://github.com/matplotlib/matplotlib/blob/699f35e4d8d424393a7038428abc573d591bd6db/lib/matplotlib/axes/_base.py#L1502-L1505

the above error is raised. 

This brings up a more general question: Wouldn't the axis and its `Scale` actually need to be asked about their data limits? Something like (in pseudo code)

    box_aspect = A * (tranformed(self.yaxis.bounds[1]) -  tranformed(self.yaxis.bounds[1]))/
                        (tranformed(self.xaxis.bounds[0]) -  tranformed(self.xaxis.bounds[0]))

such that other / custom scales can have an aspect defined?

Related issue: https://github.com/matplotlib/matplotlib/issues/8878

**Matplotlib version**

  * Matplotlib version: master

**Comment by zc62** [comment]
Would be nice to see this implemented. I first saw #4900 and thought it was a bug, then I realized it is just not implemented for scales other than `linear` and `log`.

**Comment by QuLogic** [comment]
Closed by #14727.
